### PR TITLE
Add /blocks API

### PR DESCRIPTION
**Pull request by LLFourn**
Note that all the fields have been added yet. Just some useful ones that seemed important for bdk.

### CODE REVIEW

File: src/blocking.rs [hunk 1]
```diff
@@ -306,6 +306,19 @@ impl BlockingClient {
...
+    /// Gets some recent blocks starting at the tip or at `height` if provided.
+    ///
+    /// The number of blocks depends on the backend itself but it should be at least 10.
+    pub fn blocks(&self, height: Option<u32>) -> Result<Vec<Block>, Error> {
+        let url = match height {
+            Some(height) => format!("{}/blocks/{}", self.url, height),
+            None => format!("{}/blocks", self.url),
+        };
+
+        dbg!(self.agent.get(&url).call()?.into_string()?);
```
Comment (evanlinjin) [inline src/blocking.rs:319]:
```suggestion
```
Accidentally included.

**Review by evanlinjin — APPROVED** [review]
ACK deb2bc8026720961713b4f43d97fb0e697805732 Just a small nit.

File: src/lib.rs [hunk 2]
```diff
@@ -682,4 +682,29 @@ mod test {
...
+        let blocks_async2 = async_client.blocks(None).await.unwrap();
+        assert_eq!(blocks2, blocks_async2);
+        assert_ne!(blocks2, blocks1);
+        let blocks3 = blocking_client.blocks(Some(start_height as u32)).unwrap();
+        let blocks_async3 = async_client
+            .blocks(Some(start_height as u32))
+            .await
+            .unwrap();
+        assert_eq!(blocks3, blocks_async3);
+        assert_eq!(blocks3, blocks1);
```
Comment (tnull) [inline src/lib.rs:705]: Could check here that the returned blocks at `start_height` are actually the same we can receive via `BITCOIND.client`.

File: src/blocking.rs [hunk 3]
```diff
@@ -306,6 +306,19 @@ impl BlockingClient {
         };
         Ok(self.agent.get(&url).call()?.into_json()?)
     }
+
+    /// Gets some recent blocks starting at the tip or at `height` if provided.
+    ///
+    /// The number of blocks depends on the backend itself but it should be at least 10.
```
Comment (tnull) [inline src/blocking.rs:312]: IIUC, the call usually returns 10 entries, but we can't guarantee that, e.g., for a `start_height` < 10. So, if there indeed other backends that default to return a number other than 10, maybe just leave this line out?

Comment (LLFourn) [reply src/blocking.rs]: `mempool.space` returns 15. I could always truncate to 10 if that makes things better maybe? For now I've just expanded the comment to explain that mempool returns a different value.

File: src/api.rs [hunk 4]
```diff
@@ -73,12 +73,22 @@ pub struct Tx {
...
 
-#[derive(Deserialize, Clone, Debug)]
+#[derive(Deserialize, Clone, Debug, PartialEq)]
 pub struct BlockTime {
     pub timestamp: u64,
     pub height: u32,
 }
 
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct Block {
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply src/api.rs]: I renamed it to `BlockSummary`.

File: src/async.rs [hunk 5]
```diff
@@ -272,4 +272,23 @@ impl AsyncClient {
             .json::<HashMap<String, f64>>()
             .await?)
     }
+
+    /// Gets some recent blocks starting at the tip or at `height` if provided.
+    ///
+    /// The number of blocks depends on the backend itself but it should be at least 10.
+    pub async fn blocks(&self, height: Option<u32>) -> Result<Vec<Block>, Error> {
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply src/async.rs]: Renamed and improved doc comment.

**Comment by LLFourn** [comment]
thanks @tnull I addressed all comments I believe.

File: src/async.rs [hunk 6]
```diff
@@ -275,8 +275,8 @@ impl AsyncClient {
 
     /// Gets some recent blocks starting at the tip or at `height` if provided.
     ///
-    /// The number of blocks depends on the backend itself but it should be at least 10.
-    pub async fn blocks(&self, height: Option<u32>) -> Result<Vec<Block>, Error> {
+    /// The max number of blocks returned depends on the backend itself, esplora returns `10` while [mempool.space] returns `15`.
```
Comment (tnull) [inline src/async.rs:278]:
```suggestion
    /// The maximum number of blocks returned depends on the backend itself, e.g., Esplora returns `10` while [mempool.space](https://mempool.space) returns `15`.
```

**Review by tnull — APPROVED** [review]
One tiny nit, otherwise LGTM after `cargo fmt`.

File: src/lib.rs (hunk 2 above)
Comment (vladimirfomene) [reply src/lib.rs:707]: I haven't seen this addressed.

File: src/lib.rs [hunk 7]
```diff
@@ -685,26 +685,28 @@ mod test {
...
+            .get_blocks(Some(start_height as u32))
             .await
             .unwrap();
         assert_eq!(blocks3, blocks_async3);
+        assert_eq!(blocks3[0].time.height, start_height as u32);
         assert_eq!(blocks3, blocks1);
-        let blocks_genesis = blocking_client.blocks(Some(0)).unwrap();
-        let blocks_genesis_async = blocking_client.blocks(Some(0)).unwrap();
+        let blocks_genesis = blocking_client.get_blocks(Some(0)).unwrap();
+        let blocks_genesis_async = blocking_client.get_blocks(Some(0)).unwrap();
```
Comment (vladimirfomene) [inline src/lib.rs:709]: Shouldn't this be using the `async_client`

**Comment by tnull** [comment]
LGTM, feel free to squash.

File: src/api.rs [hunk 8]
```diff
@@ -73,12 +73,22 @@ pub struct Tx {
...
 
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct BlockSummary {
+    pub id: BlockHash,
+    #[serde(flatten)]
+    pub time: BlockTime,
+    /// Hash of the previous block, will be `None` for the genesis block.
+    pub previousblockhash: Option<bitcoin::BlockHash>,
+    pub merkle_root: bitcoin::TxMerkleNode,
+}
```
Comment (rajarshimaitra) [inline src/api.rs:90]: To recreate `bdk_core::BlockId` from this we also need the  block height in the summary..

**Review by rajarshimaitra — CHANGES_REQUESTED** [review]

File: src/api.rs (hunk 8 above)
Comment (LLFourn) [reply src/api.rs:90]: It's in the `time` field.

**Comment by LLFourn** [comment]
I squashed the commits. Sorry about the epic wait on this one.

**Comment by LLFourn** [comment]
...the CI seems to be stuck.

**Comment by vladimirfomene** [comment]
Hi @LLFourn, can you rebase the changes on master and force push? That will re-run the CI jobs and I'm not able to re-run them from my end.

**Comment by notmandatory** [comment]
This will need to be rebased once #31 goes in, I think it will fix the test reliability.

**Pull request merged**